### PR TITLE
luci-app-softether: add basic softether status page

### PR DESCRIPTION
**Pull request by berenyibalazs**
This adds a basic SoftEther VPN client status page.
I have tested/using it on 19.07.4 mir3g.

### CODE REVIEW

**Review by feckert — CHANGES_REQUESTED** [review]
Please change to tabs.
Missing new line for luci-app-softether.json

**Comment by berenyibalazs** [comment]
Hi @feckert,

thanks for the comments!
What do you mean? Like on admin/system/system the tabs for Genereal/Time/Logging? For each VPN connection(account)?
I made it similar to the WireGuard status page.

BR,
Balazs.

File: applications/luci-app-softether/root/usr/share/luci/menu.d/luci-app-softether.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+        "admin/status/softether": {
+                "title": "SoftEther Status",
+                "order": 94,
+                "action": {
+                        "type": "template",
+                        "path": "softether"
+                }
+        }
+}
```
Comment (feckert) [inline applications/luci-app-softether/root/usr/share/luci/menu.d/luci-app-softether.json:10]: There must be a new line

File: applications/luci-app-softether/luasrc/view/softether.htm [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+  for acc_name, acc_properties in pairs(accounts) do
+    dump_account(acc_name, acc_properties)
+  end
+  luci.http.write("</div>") -- table
+  luci.http.write("</div>") -- section-node
+  luci.http.write("</div>") -- section
+
+ %>
+<%+footer%>
```
Comment (feckert) [inline applications/luci-app-softether/luasrc/view/softether.htm:74]: This file should be converted to tab.
Now use spaces at the beginning of the lines.
Tabs saves memory here if it is not minified

File: applications/luci-app-softether/Makefile [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+LUCI_TITLE:=SoftEther Status
+LUCI_DEPENDS:=+softethervpn5-client
+LUCI_PKGARCH:=all
+
+PKG_LICENSE:=Apache-2.0
+
+include ../../luci.mk
+
+# call BuildPackage - OpenWrt buildroot signature
```
Comment (feckert) [inline applications/luci-app-softether/Makefile:15]: Can you add pictures in the overview of the pr, how that looks?

Comment (feckert) [reply applications/luci-app-softether/Makefile:15]: Also please change the commit message to `luci-app-softether: add basic softether status page`.

**Comment by berenyibalazs** [comment]
Hi @feckert
hope i fixed everything.
Thanks, Balazs.

luci-admin-status-softether

File: applications/luci-app-softether/root/usr/share/luci/menu.d/luci-app-softether.json (hunk 1 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/root/usr/share/luci/menu.d/luci-app-softether.json]: Done.

File: applications/luci-app-softether/luasrc/view/softether.htm (hunk 2 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/luasrc/view/softether.htm]: Indentation is now donw with tabs.

File: applications/luci-app-softether/Makefile (hunk 3 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/Makefile:15]: done

File: applications/luci-app-softether/luasrc/view/softether.htm [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	luci.http.write("<div class=\"cbi-section-node\">")
+	luci.http.write("<div class=\"table cbi-section-table\">")
+	for acc_name, acc_properties in pairs(accounts) do
+		dump_account(acc_name, acc_properties)
+	end
+	luci.http.write("</div>") -- table
+	luci.http.write("</div>") -- section-node
+	luci.http.write("</div>") -- section
+%>
+<%+footer%>
```
Comment (feckert) [inline applications/luci-app-softether/luasrc/view/softether.htm:70]: Add new line

File: applications/luci-app-softether/Makefile [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#
```
Comment (feckert) [inline applications/luci-app-softether/Makefile:1]: I can not see any acl file under `root/usr/share/rpcd/acl.d/luci-app-softether.json`
I have not tested it now in master. But if this file is not there then the menu entry should not be displaye.

**Review by feckert — CHANGES_REQUESTED** [review]

File: applications/luci-app-softether/luasrc/view/softether.htm (hunk 4 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/luasrc/view/softether.htm]: Done.

File: applications/luci-app-softether/Makefile (hunk 5 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/Makefile:1]: I checked on snapshot r15252. It is the other way around. The menu item was shown even on logon screen.
Added a generic ACL reference like in other status page.
Done.

**Comment by berenyibalazs** [comment]
@feckert Is it possible to merge this PR to 19.07 branch?
Then i will make a javascript based one for master. On the other hand do javascript based ones working on 19.07?

**Comment by berenyibalazs** [comment]
Managed to modernize to Javascript.

File: applications/luci-app-softether/root/usr/share/rpcd/acl.d/luci-app-softether.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
+	"luci-app-softether": {
+		"description": "Grant access to softether management",
+		"read": {
+			"cgi-io": [ "exec" ],
+			"file": {
+				"/usr/libexec/vpncmd-call account-list": [ "exec" ]
+			}
+		},
+		"write": {
```
Comment (feckert) [inline applications/luci-app-softether/root/usr/share/rpcd/acl.d/luci-app-softether.json:10]: I don't know it exactly, but is that right with the `write`? Is not a `read` sufficient?

File: applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+				E('div', {'class':'td', 'style': 'width: 20%;vertical-align:top;'}, [
+					E('strong', {}, ['Account:']),
+					account.name
+				]),
+				E('div', {'class':'td'}, properties)
+			]);
+	return row;
+}
+
+return view.extend({
```
Comment (feckert) [inline applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js:107]: We would also put all the functions `enderAccountRow`, `updateAccountTable` into the new object like in:
* https://github.com/openwrt/luci/blob/master/modules/luci-mod-status/htdocs/luci-static/resources/view/status/processes.js 
* https://github.com/openwrt/luci/blob/master/modules/luci-mod-status/htdocs/luci-static/resources/view/status/iptables.js

**Review by feckert — CHANGES_REQUESTED** [review]

File: applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	load: function() {},
```
[the remark this answers is not in the document]
Comment (berenyibalazs) [reply applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js]: Done.

File: applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	render: function() {
```
[the remark this answers is not in the document]
Comment (berenyibalazs) [reply applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js:105]: Done.

File: applications/luci-app-softether/root/usr/share/rpcd/acl.d/luci-app-softether.json (hunk 6 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/root/usr/share/rpcd/acl.d/luci-app-softether.json]: Moved to "read".

File: applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js (hunk 7 above)
Comment (berenyibalazs) [reply applications/luci-app-softether/htdocs/luci-static/resources/view/softether-status.js]: Done.

**Review by feckert — APPROVED** [review]
I don't use softether, but the source looks good so far.

**Comment by feckert** [comment]
Thanks :+1

**Pull request merged**